feat: add generator for capstone

cmd05 opened the pull request:
Fixes #1146. Capstone is using LLVM Tablegen for its RISC-V dependencies.  However, some sections of the code can be automatically generated. 

The script `generate_csr_switch.py` generates the list of CSRs in the function [getCSRSystemRegisterName](https://github.com/capstone-engine/capstone/blob/next/arch/RISCV/RISCVInstPrinter.c#L301-L766) in `arch/RISCV/RISCVInstPrinter.c`.

Output:
```
root@edb48186766d:/workspaces/riscv-unified-db# ./do gen:capstone
Using devcontainer environment
Recreating bundle config...
Running with 1 job(s)
/workspaces/riscv-unified-db/.home/.venv/bin/python3 /workspaces/riscv-unified-db/backends/generators/capstone/generate_csr_switch.py --csr-dir=/workspaces/riscv-unified-db/gen/resolved_spec/_/csr --arch=BOTH --output=/workspaces/riscv-unified-db/gen/capstone/csr_switch.c
INFO:: Searching for CSR files in /workspaces/riscv-unified-db/gen/resolved_spec/_/csr for target architecture BOTH
INFO:: Found 382 CSR definitions in 382 files
INFO:: Added 382 CSRs to the output
Generated: /workspaces/riscv-unified-db/gen/capstone/csr_switch.c
```

@ThinkOpenly @AFOliveira

dhower-qc reviewed (changes requested):
Cool, thanks. A few requests:

1. There is a background effort to try to move what is in `backends` to `tools`. Since this is a new tool, could you please move it there? Something like `tools/python-packages/udb-capstone`.
2. Give yourself credit! Add copyright/license information to the new files. See other files in the repo for an example that is Reuse-compatible.
3. Add a regression test so that we can make sure this continues to work going forward.

ThinkOpenly commented:
> 2. Give yourself credit! Add copyright/license information to the new files. See other files in the repo for an example that is Reuse-compatible.

Indeed!

> 3. Add a regression test so that we can make sure this continues to work going forward.

Thinking about this, it's not straightforward, given you are generating the contents of an internal function. Brainstorming a bit... a few ideas:
- The function is very basic. You could download a (known version of) the source, extract the function and build it with a test routine that invokes it with each of the possible parameter values it expects. Then, build the same test routine with your generated function body, and verify the results are the same.
- You could download and build Capstone and write a test that uses Capstone to exercise the function in question, then replace the function body with your generated content and compare.
- You could also create a "golden reference" output based on the Capstone source, and compare your test output with that.

The hard part is the extract/replace.

ThinkOpenly commented:
Here's a basic script which exercises the function in Capstone:
```
#!/usr/bin/env python3
from capstone import *
md = Cs(CS_ARCH_RISCV,CS_MODE_32)
for CSR in range(2 ** 12 - 1):
    csrr_hex = f"{CSR:03x}020f3"
    # byte swap
    csrr = csrr_hex[6] + csrr_hex[7] + csrr_hex[4] + csrr_hex[5] + csrr_hex[2] + csrr_hex[3] + csrr_hex[0] + csrr_hex[1]
    csrr_bytes = bytes.fromhex(csrr)
    for i in md.disasm(csrr_bytes, 0x1000):
        print("%4d(%03x) %s\t%s" % (CSR, CSR, i.mnemonic, i.op_str))
```

Note that not all of the range of CSR addresses are mapped to a string, obviously, and some of the CSRs are associated with pseudoinstructions, and Capstone will emit those instead, so, you'll something like:
```
   0(000) csrr  ra, ustatus
   1(001) frflags       ra
   2(002) frrm  ra
   3(003) frcsr ra
   4(004) csrr  ra, uie
   5(005) csrr  ra, utvec
   6(006) csrr  ra, 6
   7(007) csrr  ra, 7
   8(008) csrr  ra, 8
   9(009) csrr  ra, 9
```

cmd05 commented:
@ThinkOpenly I have updated the code and added the regression test. I have tested the code with changes from #1236 and #1251. The regression test will currently fail because the below CSRs currently have incorrect addresses in the latest Capstone python package (fixed in this commit https://github.com/capstone-engine/capstone/commit/232e1ae3590c643111186f04c3d3e39b4c78cd26)
```
{('pmpaddr13', 958), ('pmpaddr14', 957)}
```

cmd05 commented:
I have addressed the comments above. The regression test for capstone is working on the CI and will pass when #1236 is merged

ThinkOpenly reviewed (changes requested):
If you aren't willing or able to make this changes, let us know.

ThinkOpenly commented:
> failing on the CI for regress-complete and regress-pre-commit.

At least some of the regress-pre-commit failures are fairly trivial. Are you able address the issues reported?

> The capstone regression test will pass since `sie` has been added now.

It's not, currently. Could you rebase this PR to pick that up?

cmd05 committed:
feat: add generator for capstone
